File: Traversal Algorithms/Graph Traversals/BFS/C++/BFS.cpp
// Directed Graph

#include <iostream>
#include <list>
#include <queue>
#include <set>
#include <string>
#include <unordered_map>

class Graph {
   private:
    class Node {
       public:
        std::string label;
        Node(const std::string& label) : label(label) {}
    };

    // use map for fast lookup
    std::unordered_map<std::string, Node*> nodes;

    // list of lists of adjacent node
    // eg. Node A = [B,C,D] where, B,C,D are adjacent to A
    std::unordered_map<Node*, std::list<Node*>> adjacencyList;

   public:
    void addNode(const std::string&);
    void addEdge(const std::string&, const std::string&);

    // Breadth First Traversal
    void breadthFirstTraversal(const std::string&);
};

void Graph::addNode(const std::string& label) {
    Node* new_node = new Node(label);
    nodes.emplace(label, new_node);
}

void Graph::addEdge(const std::string& from, const std::string& to) {
    if (nodes.find(from) == nodes.end())
        throw std::runtime_error("Error: " + from + " node not found");
    if (nodes.find(to) == nodes.end())
        throw std::runtime_error("Error: " + to + " node not found");

    // add "to node" to "from node" list as its adjacent node
    adjacencyList[nodes[from]].push_back(nodes[to]);
}

void Graph::breadthFirstTraversal(const std::string& root) {
    // validate the root node
    if (nodes.find(root) == nodes.end())
        throw std::runtime_error("Error: " + root + " node not found");

    // set of visited nodes
    std::set<Node*> visited;

    // create a queue frontier and push root node
    std::queue<Node*> frontier;
    frontier.push(nodes[root]);

    while (!frontier.empty()) {
        // pop one node from the front of the queue
        Node* current = frontier.front();
        frontier.pop();

        // continue if the current node is already visited
        if (visited.find(current) != visited.end())
            continue;

        // visit the current node and add it to the set of visited node
        std::cout << current->label << std::endl;
        visited.insert(current);

        // push all of current's unvisited neighbour into the queue
        for (auto&& neighbour : adjacencyList[current]) {
            if (visited.find(neighbour) == visited.end())
                frontier.push(neighbour);
        }
    }
}

/*
    To add a node  : graph.addNode("A");
    To add an edge : graph.addEdge("A", "B");

    Sample input graph:

        ┌──A──┐
        │     │
        B  ┌──C──┐
           │     │
        ┌──D──┐  E
        │     │
        F     G

    Sample Output:
        A
        B
        C
        D
        E
        F
        G

*/

File: Traversal Algorithms/Graph Traversals/BFS/JAVA/BFS.java
import java.util.*;

public class BFS{

    private static int V;   // No. of vertices
    private static LinkedList<Integer> adj[]; //Adjacency Lists
 
    // Constructor
    BFS(int v)
    {
        V = v;
        adj = new LinkedList[v];
        for (int i = 0; i < v; ++i){
            adj[i] = new LinkedList();
        }
    }

    // Function to add an edge into the graph
    void addEdge(int v,int w)
    {
        adj[v].add(w);
    }
 
    // prints BFS traversal from a given source s
    public static void BFS(int s)
    {
        boolean visited[] = new boolean[V]; // Mark all the vertices as not visited
        
        LinkedList<Integer> queue = new LinkedList<Integer>();  
 
        visited[s]=true;    // Mark the current node as visited and enqueue it
        queue.add(s);
 
        while (queue.size() != 0)
        {
            s = queue.poll();
            System.out.print(s+" ");    // Dequeue a vertex from queue and print it
                       
            Iterator<Integer> i = adj[s].listIterator();  // Get all adjacent vertices of the dequeued vertex s
            while (i.hasNext())
            {
                int n = i.next();
                if (!visited[n])
                {
                    visited[n] = true;  // If a adjacent has not been visited, then mark it visited and enqueue it
                    queue.add(n);
                }
            }
        }
    }

}

/*
    * Sample Input: (0,1) (0,2) (1,2) (2,0) (2,3) (3,3)
    * Sample Output: 2 0 3 1
    * Time Complexity in Best Case: O(V+E)
	* Time Complexity in Average Case: O(V+E)
	* Time Complexity in Worst Case: O(V+E)
	* Space Complexity in Worst Case: O(V) 
*/